### PR TITLE
DM-29543: Gen3 refcat converter

### PR DESCRIPTION
**Pull request by parejkoj**

### CODE REVIEW

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (erykoff) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:141]: Another question: if you're suggesting `direct` then the refcat files have to be in the location where they're supposed to be, right?  For the shared repos, should this be run in the `refcats` subdirectory (with write access)?  I think this needs more detail on what should be done by the next person who comes along and does DR3 or DR4, etc.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (erykoff) [inline doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:93]: I don't think this is correct, and this confused me.  These columns are special-cased by the gaia manager, and do not go through the regular code (which would be wrong because the gaia magnitude columns are Vega, while the regular code assumes AB).  Thankfully, I think that the code is correct but this config is confusing because it's not used as best as I can tell.

**Review by erykoff — CHANGES_REQUESTED** [review]
Looks fine overall, but see comments below.

My biggest comment is a separate use case that might be out of scope for this ticket but want to make sure this is compatible... what to do if I have either (a) a big in memory catalog, or (b) an even bigger catalog that is a bunch of non-csv files (regular fits, afw catalogs, parquet, etc) that I would like to turn into a refcat.  My current reading of the docs/code is that I have to write my own sharding code, schema-writing code, and an ingest table file.  In which case the schema and ingest table code should be made into static or free functions, or there should be a way to get into this code without having csv files.  (Some of this might be misreading the code; it's a bit late but I would have suggested abandoning the gen2 code rather than refactoring it even at the expense of code duplication).

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (timj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:141]: Yes, direct means you are promising they are in their final archival location and won't be moved/deleted. This is essentially the same as using a symlink of course -- the same promise is required.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:22]: Sorry, which filename is mispelled? The :ref: link there is to a section in this document.

Comment (erykoff) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:22]: "suitable filaname->htm7-index"

Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:22]: The word "filename", not some filename somewhere... right.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:34]: The fluxes and flux errors they supply are e-/s, which are not something we can easily use, hence the special-cased `ConvertGaiaManager`.

You are correct that the Gaia-supplied magnitudes are Vega, and they aren't actually used. `mag_column_list` is a required config field and is used to populate the output schema, which is why it exists here in the docs. I'll update the comment attached to it.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:45]: I've added an additional note about the `ReadFitsCatalogTask`.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:141]: There's a sentence just below this that "These commands should finish in a short amount of time, logging a message about how many files were ingested." - I haven't timed it on the full Gaia catalog with 128,000 files, but I wouldn't expect it to take more than a couple minutes at most with the `direct` mode.

Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:141]: As to the question of what to do for e.g. DR3 and shared repos, we need someone on the survey policy side to update the Common Dataset Policy document. I filed DM-31704 on this topic, and just added you to it.
https://developer.lsst.io/services/datasets.html#reference-catalogs

File: python/lsst/meas/algorithms/convertReferenceCatalog.py
Comment (parejkoj) [reply python/lsst/meas/algorithms/convertReferenceCatalog.py:70]: This step takes so little time compared with the rest of the conversion, I'm not sure it's worth the bother.

File: python/lsst/meas/algorithms/ingestIndexReferenceTask.py
Comment (parejkoj) [reply python/lsst/meas/algorithms/ingestIndexReferenceTask.py:113]: Yeah, I've been torn on how to handle this one. We'll need people to override it (to completely get rid of `cal_ref_cat`!), but if we remove the default right now, that would actually break some gen2 stuff. The deprecation warning here is rather annoying; I could create a ticket blocked on the gen2 removal ticket (DM-20572) to remove this default during or after that process, since it really only matters in gen3 for refcat creation, which is something that very few users actually do.

File: python/lsst/meas/algorithms/ingestIndexReferenceTask.py
Comment (parejkoj) [reply python/lsst/meas/algorithms/ingestIndexReferenceTask.py:498]: Refactoring it was very helpful for me in finding all the places (actually only a few!) that had gen3 butler interactions, and turning them into simple file operations. Once gen2 really is gone, we can think about refactoring to merge the base class back into here.

Although, thinking about your comment above about converting in-memory catalogs, maybe the baseclass would be useful for building something like that.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (parejkoj) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:93]: I updated the comment on this required column to make it more clear.

**Comment by parejkoj** [comment]
Your suggestion about converting/sharding an in-memory catalog is a good one for another ticket: I think we'd want to carefully consider what such an input catalog would look like, and whether the code that produces it (I'm guessing fgcmcal, in this case?) can get the in-memory representation right, so it only has to be sharded, not have fields converted.

If you have a parquet table, you'd have to write a `readParquetCatalogTask`, which would be pretty trivial, I think.

We have a `readFitsCatalogTask` (I updated the docs with a link to it), which was made to work here. It would also work for converting an afw catalog; I don't know whether that's a completely trivial thing (requiring only sharding), given differing column naming conventions in SourceCatalogs vs. refcats.

File: doc/lsst.meas.algorithms/creating-a-reference-catalog.rst
Comment (erykoff) [reply doc/lsst.meas.algorithms/creating-a-reference-catalog.rst:34]: I think I had too many comments on this which led to some confusion.  I understand why `ConvertGaiaManager` is necessary.  I don't think the comment that "Gaia DR2 does not supply magnitude errors" is relevant, since we don't use the magnitudes either.  Maybe "(for example, with Gaia DR2 we convert native fluxes and errors in e-/s to AB magnitudes".

File: python/lsst/meas/algorithms/ingestIndexReferenceTask.py
Comment (erykoff) [reply python/lsst/meas/algorithms/ingestIndexReferenceTask.py:113]: My issue here is that the deprecation warning will come when people use this _correctly_ (setting an override) and there won't be a warning when people don't override it (which is when we'd need a warning).  So the deprecation code that we have just doesn't work.  I think we just need to file a ticket with a TODO here.

File: python/lsst/meas/algorithms/ingestIndexReferenceTask.py
Comment (erykoff) [reply python/lsst/meas/algorithms/ingestIndexReferenceTask.py:498]: That would be useful!  I want to make sure that this is put together in a way that the key parts are accessible without running the full task.

**Review by erykoff — APPROVED** [review]

**Pull request merged**
